### PR TITLE
BUG: Series.any() and .all() don't return bool values if dtype=object

### LINKED ISSUE #12863: Series.any() and .all() don't return bool values if dtype=object
#### Code Sample, a copy-pastable example if possible

```
s = pd.Series(index=range(5), data=['a', 'b', 'c', 'd', 'e'], dtype=object)
print [s.any(), s.all()]
```
#### Expected Output

`[True, True]`
#### Actual output

`['a', 'e']`
#### output of `pd.show_versions()`

```
INSTALLED VERSIONS
------------------
commit: None
python: 2.7.11.final.0
python-bits: 64
OS: Darwin
OS-release: 15.0.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: None
nose: 1.3.7
pip: 8.1.1
setuptools: None
Cython: None
numpy: 1.10.1
scipy: 0.16.1
statsmodels: None
IPython: 4.0.0
sphinx: 1.3.3
patsy: None
dateutil: 2.4.2
pytz: 2015.7
blosc: None
bottleneck: None
tables: None
numexpr: None
matplotlib: 1.4.3
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: None
httplib2: None
apiclient: None
sqlalchemy: 1.0.11
pymysql: None
psycopg2: None
Jinja2: None
```

### PR DESCRIPTION
**Pull request by ShaharNaveh**
- [x] closes #12863
- [ ] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/core/nanops.py [hunk 1]
```diff
@@ -453,7 +457,11 @@ def nanall(values, axis=None, skipna: bool = True, mask=None):
     >>> nanops.nanall(s)
     False
     """
-    values, _, _, _, _ = _get_values(values, skipna, fill_value=True, mask=mask)
+    values, _, dtype, _, _ = _get_values(values, skipna, fill_value=True, mask=mask)
+
+    if is_object_dtype(dtype):
+        return all(values)
```
Comment (jbrockmendel) [inline pandas/core/nanops.py]: can you use `values.all()` here?  that is usually more performant

Comment (jbrockmendel) [reply pandas/core/nanops.py]: (might also be relevant to the CI failures)

File: pandas/core/nanops.py [hunk 2]
```diff
@@ -421,7 +421,11 @@ def nanany(values, axis=None, skipna: bool = True, mask=None):
     >>> nanops.nanany(s)
     False
     """
-    values, _, _, _, _ = _get_values(values, skipna, fill_value=False, mask=mask)
+    values, _, dtype, _, _ = _get_values(values, skipna, fill_value=False, mask=mask)
+
+    if is_object_dtype(dtype) and axis is None:
```
Comment (jreback) [inline pandas/core/nanops.py]: can you add a ref to this issue here (and in nanall where you added code): https://github.com/numpy/numpy/issues/4352, and a small comment ( like we are doing the right thing as numpy doesn't handle any/all on object arrays correct)

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm. can you add a whatsnew note. i think this deserves a small subsection (even though its a bug fix), you can put in api-breaking changes, just show the before and after (model after some existing changes like this)

File: pandas/core/nanops.py (hunk 1 above)
Comment (MomIsBestFriend) [reply pandas/core/nanops.py]: > can you use `values.all()` here? that is usually more performant

That's exactly the bug!

Comment (jbrockmendel) [reply pandas/core/nanops.py]: I see.  what about `np.all(values)` and `np.any(values)`?

Comment (MomIsBestFriend) [reply pandas/core/nanops.py]: > I see. what about `np.all(values)` and `np.any(values)`?

Seem to work, I will mark this as resolve if all the test cases passes.

Comment (MomIsBestFriend) [reply pandas/core/nanops.py]: @jbrockmendel It doesn't work

**Comment by MomIsBestFriend** [comment]
The only implementation I can think of right now, is to create a function to iterate over the values of the np.ndarray as return ```False``` if it found one of:
```python
(np.nan, None)
```

tl;dr
Implement ```any``` and ```all``` methods.

File: pandas/core/nanops.py [hunk 3]
```diff
@@ -421,7 +421,16 @@ def nanany(values, axis=None, skipna: bool = True, mask=None):
     >>> nanops.nanany(s)
     False
     """
-    values, _, _, _, _ = _get_values(values, skipna, fill_value=False, mask=mask)
+    values, _, dtype, _, _ = _get_values(values, skipna, fill_value=False, mask=mask)
+
+    # GH #12863
+    # Checking if the `axis` is None because numpy
+    # doesn't handle ``any`` and ``all`` on
+    # object arrays correclty. see
```
Comment (jbrockmendel) [inline pandas/core/nanops.py]: correclty -> correctly.  small preference for this to go inside the "if", not sure how widely shared that is

File: pandas/tests/reductions/test_reductions.py [hunk 4]
```diff
@@ -811,19 +811,33 @@ def test_all_any(self):
...
 
     def test_all_any_params(self):
         # Check skipna, with implicit 'object' dtype.
         s1 = Series([np.nan, True])
         s2 = Series([np.nan, False])
-        assert s1.all(skipna=False)  # nan && True => True
+
         assert s1.all(skipna=True)
-        assert np.isnan(s2.any(skipna=False))  # nan || False => nan
+        assert s1.any(skipna=True)
```
Comment (jbrockmendel) [inline pandas/tests/reductions/test_reductions.py]: pls add the GH ref here too

File: pandas/tests/reductions/test_reductions.py [hunk 5]
```diff
@@ -811,19 +811,33 @@ def test_all_any(self):
         assert not bool_series.all()
         assert bool_series.any()
 
-        # Alternative types, with implicit 'object' dtype.
-        s = Series(["abc", True])
-        assert "abc" == s.any()  # 'abc' || True => 'abc'
+        # GH 12863
```
Comment (jbrockmendel) [inline pandas/tests/reductions/test_reductions.py]: might as well make this a separate test test_all_any_object_dtype

**Comment by jbrockmendel** [comment]
Does this bugfix affect DataFrame at all?  if so, test for that?

optional: could test the affected nanops functions directly in test_nanops.

**Comment by jbrockmendel** [comment]
what case is breaking that necessitates the brute force patch?

**Comment by MomIsBestFriend** [comment]
> what case is breaking that necessitates the brute force patch?

The original test case (from #12863)
```python
pd.Series(index=range(5), data=['a', 'b', 'c', 'd', 'e'])
```

**Comment by MomIsBestFriend** [comment]
I can't figure it out, when the original case is working, every test is failing.
I can xfail them all, but it's just seems wrong, that so many fails.

File: pandas/core/nanops.py [hunk 6]
```diff
@@ -421,8 +421,44 @@ def nanany(values, axis=None, skipna: bool = True, mask=None):
...
+    # Checking if the `axis` is None because numpy
+    # doesn't handle ``any`` and ``all`` on
+    # object arrays correctly. see
+    # https://github.com/numpy/numpy/issues/4352
+
+    # TODO: Find a less code-smelly way of doing this
+    if is_object_dtype(dtype) and axis is None:
+        output = values.any()
+    else:
+        output = values.any(axis)
```
Comment (jreback) [inline pandas/core/nanops.py]: you can just return directly from here.

File: pandas/core/nanops.py [hunk 7]
```diff
@@ -421,8 +421,44 @@ def nanany(values, axis=None, skipna: bool = True, mask=None):
...
+
+    if isinstance(output, bool):
```
Comment (jreback) [inline pandas/core/nanops.py]: when is this NOT true?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/nanops.py (hunk 7 above)
Comment (MomIsBestFriend) [reply pandas/core/nanops.py]: That's exactly the bug, for example (from the original issue).

master:
```python
>>> import pandas as pd
>>> s = pd.Series(index=range(5), data=['a', 'b', 'c', 'd', 'e'], dtype=object)
>>> s.any()
'a'
>>> s.all()
'e'
```

This branch (with brute force function):
```python
>>> import pandas as pd
>>> s = pd.Series(index=range(5), data=['a', 'b', 'c', 'd', 'e'], dtype=object)
>>> s.any()
True
>>> s.all()
True
```

**Comment by jreback** [comment]
can you merge master and we'll take a look

**Comment by MomIsBestFriend** [comment]
> can you merge master and we'll take a look

I have merged master @jreback

**Comment by jreback** [comment]
@MomIsBestFriend so the tests are failing, ping when green.

**Comment by MomIsBestFriend** [comment]
> @MomIsBestFriend so the tests are failing, ping when green.

@jreback I'm really not sure what to do, I see pretty much every where the error message of:

```python
AttributeError: type object 'bool' has no attribute 'any'
```

No idea on where to even start. should I change the ```nanany```/```nanall```?

or change about 30 - 50 tests, it's just seems wrong to do so.

**Comment by jreback** [comment]
> > @MomIsBestFriend so the tests are failing, ping when green.
> 
> @jreback I'm really not sure what to do, I see pretty much everywhere the error message of:
> 
> ```python
> AttributeError: type object 'bool' has no attribute 'any'
> ```
> 
> No idea on where to even start. should I change the `nanany`/`nanall` functions?
> 
> or change about 30 - 50 tests, it's just seems wrong to do so.

try using np.any

**Comment by MomIsBestFriend** [comment]
> > > @MomIsBestFriend so the tests are failing, ping when green.
> > 
> > 
> > @jreback I'm really not sure what to do, I see pretty much everywhere the error message of:
> > ```python
> > AttributeError: type object 'bool' has no attribute 'any'
> > ```
> > 
> > 
> > No idea on where to even start. should I change the `nanany`/`nanall` functions?
> > or change about 30 - 50 tests, it's just seems wrong to do so.
> 
> try using np.any

This also doesn't work.

**Comment by jbrockmendel** [comment]
> This also doesn't work.

@MomIsBestFriend now that #29847 has been merged, I'm planning to do a significant cleanup of DataFrame._reduce and core.nanops.  So I think the path forward is for you and I to trade tasks.  Up for something new?

**Comment by MomIsBestFriend** [comment]
> Up for something new?

Yes just take it away from me :laughing:

**Comment by MomIsBestFriend** [comment]
Closing this, because this I don't know how to fix it without breaking everything else.

**Comment by gimseng** [comment]
Did the issue eventually got fixed? Can someone tldr the thread? Thanks !
